packages/drivers/odsp-driver/src/odspDocumentService.ts

```diff
@@ -304,9 +297,67 @@ export class OdspDocumentService implements IDocumentService {
         });
     }
 
+    private clearJoinSessionTimer() {
+        if (this.joinSessionRefreshTimer !== undefined) {
+            clearTimeout(this.joinSessionRefreshTimer);
+            this.joinSessionRefreshTimer = undefined;
+        }
+    }
+
+    private async refreshSessionPeriodically(requestWebsocketTokenFromJoinSession: boolean) {
+        for (let i = 0;;i++) {
+            // If the refresh time is not defined, then let the session expire.
+            if (this.joinSessionRefreshTime === undefined || this.joinSessionRefreshTime <= 0) {
+                this.mc.logger.sendErrorEvent({
+                    eventName: "JoinSessionRefreshTimeNotSet",
+                    joinSessionRefreshTime: this.joinSessionRefreshTime,
+                    timerRunning: this.joinSessionRefreshTimer !== undefined,
+                });
+                break;
+            }
+            await new Promise<void>((resolve) => {
+                assert(this.joinSessionRefreshTime !== undefined, "Join session Refresh time should be defined");
+                this.joinSessionRefreshTimer = setTimeout(async () => {
+                    await getWithRetryForTokenRefresh(async (options) => {
+                        await this.joinSession(requestWebsocketTokenFromJoinSession, options, true);
+                        resolve();
+                    });
+                }, this.joinSessionRefreshTime - Date.now());
```

Comment:
Any reason for this.joinSessionRefreshTime to be timestamp, not delta? Due to math back and forward. the value we get here can be negative (if, for example, JS engine was throttled for a minute). Also your check above for this.joinSessionRefreshTime <= 0 does not make much sense given it's absolute time, not delta. Consider using delta.

Reply:
That said, maybe it's fine (as long as we do some small adjustments here) - please see other comments

Reply:
Using the delta now instead of timestamp.